---
title: "R Notebook"
output: html_notebook
---


```{r}
library(Seurat)
library(dplyr)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "~/Box Sync/laptop_folders/Willenbring/Feng Single Cell/willenbringh/FC_102418_analysis/filtered_gene_bc_matrices/mm10/")

# Examine the memory savings between regular and sparse matrices
dense.size <- object.size(x = as.matrix(x = pbmc.data))
dense.size
```

```{r}
sparse.size <- object.size(x = pbmc.data)
sparse.size
```

```{r}
dense.size/sparse.size
```

```{r}
pbmc <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 3, min.genes = 200, 
    project = "10X_PBMC")
```

```{r}
# The number of genes and UMIs (nGene and nUMI) are automatically calculated
# for every object by Seurat.  For non-UMI data, nUMI represents the sum of
# the non-normalized values within a cell We calculate the percentage of
# mitochondrial genes here and store it in percent.mito using AddMetaData.
# We use object@raw.data since this represents non-transformed and
# non-log-normalized counts The % of UMI mapping to MT-genes is a common

# scRNA-seq QC metric.
mito.genes <- grep(pattern = "^mt-", x = rownames(x = pbmc@data), value = TRUE)
percent.mito <- Matrix::colSums(pbmc@raw.data[mito.genes, ])/Matrix::colSums(pbmc@raw.data)

# AddMetaData adds columns to object@meta.data, and is a great place to
# stash QC stats
pbmc <- AddMetaData(object = pbmc, metadata = percent.mito, col.name = "percent.mito")
VlnPlot(object = pbmc, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)

```

```{r}
# GenePlot is typically used to visualize gene-gene relationships, but can
# be used for anything calculated by the object, i.e. columns in
# object@meta.data, PC scores etc.  Since there is a rare subset of cells
# with an outlier level of high mitochondrial percentage and also low UMI
# content, we filter these as well
par(mfrow = c(1, 2))
GenePlot(object = pbmc, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = pbmc, gene1 = "nUMI", gene2 = "nGene")
```


```{r}
# We filter out cells that have unique gene counts over 2,500 or less than
# 200 Note that low.thresholds and high.thresholds are used to define a
# 'gate'.  -Inf and Inf should be used if you don't want a lower or upper
# threshold.
pbmc <- FilterCells(object = pbmc, subset.names = c("nGene", "percent.mito"), 
    low.thresholds = c(200, -Inf), high.thresholds = c(8000, 0.50))
```


```{r}
pbmc <- NormalizeData(object = pbmc, normalization.method = "LogNormalize", 
    scale.factor = 10000)
pbmc <- FindVariableGenes(object = pbmc, mean.function = ExpMean, dispersion.function = LogVMR, 
    x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)
pbmc <- ScaleData(object = pbmc, vars.to.regress = c("nUMI", "percent.mito"))
pbmc <- RunPCA(object = pbmc, pc.genes = pbmc@var.genes, do.print = TRUE, pcs.print = 1:5, 
    genes.print = 5)

```

```{r}
VizPCA(object = pbmc, pcs.use = 1:12)

```

```{r}
PCAPlot(object = pbmc, dim.1 = 1, dim.2 = 2)
PCAPlot(object = pbmc, dim.1 = 1, dim.2 = 3)
PCAPlot(object = pbmc, dim.1 = 2, dim.2 = 3)
```

```{r}
pbmc <- ProjectPCA(object = pbmc, do.print = FALSE)
PCHeatmap(object = pbmc, pc.use = 1:4, cells.use = 200)
#dim(0, do.balanced = TRUE, label.columns = FALSE)
```
```{r}
marker_genes = c("Hnf4a", "Alb", "Fah", "Krt8", "Krt18", "G6pc", "Sox9", "Krt7", "Krt19", "Epcam", "Onecut1", "Afp", "Xist")
```

```{r}
vlnPlot(pbmc,c("Hnf4a", "Alb", "Fah"))

```

```{r}
pbmc <- FindClusters(object = pbmc, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
pbmc <- RunTSNE(object = pbmc, dims.use = 1:10, do.fast = TRUE)
```


```{r}
TSNEPlot(object = pbmc)
PrintFindClustersParams(object = pbmc)
VlnPlot(object = pbmc, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)
```

```{r}
#install.packages( "devtools" )
#devtools::install_github( "anders-biostat/JsRCom" )
#devtools::install_github( "anders-biostat/sleepwalk" )
library(sleepwalk)
#sleepwalk(data.frame(pbmc@dr[["tsne"]]@cell.embeddings), featureMatrices = data.matrix(t(pbmc@data)))
sleepwalk(as.matrix(pbmc@dr[["tsne"]]@cell.embeddings), featureMatrices = as.matrix(t(pbmc@data)))
```


```{r}
DotPlot(pbmc, marker_genes)
```

```{r}
#library(MAST)
#pbmc=buildClusterTree(pbmc,do.reorder = TRUE,reorder.numeric = TRUE,pcs.use = 1:11)
#plotClusterTree(pbmc)

  singler = CreateSinglerObject(pbmc.data, annot = NULL, project.name="MCA", min.genes = 0,
  technology = "10X", species = "Mouse", citation = "",
  ref.list = list(), normalize.gene.length = F, variable.genes = "de",
  fine.tune = T, do.signatures = T, clusters = NULL, do.main.types = T, 
  reduce.file.size = T, numCores = SingleR.numCores)
```

```{r}
markers <- FindMarkers(object = pbmc, ident.1 = 5)
write.table(markers, "cluster5.txt", sep="\t")
markers <- FindMarkers(object = pbmc, ident.1 = 6)
write.table(markers, "cluster6.txt", sep="\t")
markers <- FindMarkers(object = pbmc, ident.1 = 7)
write.table(markers, "cluster7.txt", sep="\t")
markers <- FindMarkers(object = pbmc, ident.1 = 8)
write.table(markers, "cluster8.txt", sep="\t")
markers <- FindMarkers(object = pbmc, ident.1 = 5:8)
write.table(markers, "cluster5thru8.txt", sep="\t")
```

```{r}

temp = (as.matrix(pbmc.data))
hist(log2(colSums(temp))+1, main="sum of counts for all cells")
hist(log2(rowSums(temp))+1, main= "sum of counts for all genes")

#which cells express Alb above a certain point
hist(log2(temp[rownames(temp)== "Alb", ]+1))
hist(log2(temp[rownames(temp)== "Epcam", ]+1))
hist(log2(temp[ rownames(temp) == "Epcam", which(temp[rownames(temp)== "Sox9", ]>4)]+1))
hist(log2(temp[rownames(temp)== "Xist", ]+1))
hist(log2(temp[ rownames(temp) == "Xist", which(temp[rownames(temp)== "Sox9", ]>4)]+1))
hist(log2(temp[rownames(temp)== "Hnf4a", ]+1))
hist(log2(temp[rownames(temp)== "Onecut1", ]+1))
hist(log2(temp[rownames(temp)== "Sox9", ]+1))

```

```{r}
library(pheatmap)
temp3 = temp[which(rownames(temp) %in% marker_genes), ]
pheatmap(log2(temp3+1), show_colnames=F)
temp6 = apply(temp, 2, function (x) (x+0.000000000000000000000001)/sum(x))
temp7 = temp6[which(rownames(temp6) %in% marker_genes), ]
pheatmap(log2(temp7), show_colnames=F)
```

genes expressing Albumin don't express epcam much
genes expressing epcam have some albumin expression
```{r}
alb_exp = temp6[, which(temp6 [which(rownames(temp6)=="Alb"),]>0)]
alb_exp_markers = alb_exp[which(rownames(alb_exp) %in% marker_genes), ]
pheatmap(log2(alb_exp_markers), show_colnames=F)
epcam_exp = temp6[, which(temp6 [which(rownames(temp6)=="Epcam"),]>0)]
epcam_exp_markers = epcam_exp[which(rownames(epcam_exp) %in% marker_genes), ]
pheatmap(log2(epcam_exp_markers), show_colnames=F)
```


```{r}
singler = SingleR(method = "single", sc_data, ref_data, types, clusters = NULL,
  genes = "de", quantile.use = 0.8, p.threshold = 0.05,
  fine.tune = TRUE, fine.tune.thres = 0.05, sd.thres = 1,
  do.pvals = T, numCores = SingleR.numCores)
```

