---
title: "2018_09_07_Willenbring_motif_enrichment"
output: html_document
---

setup packages
```{r }
library(plyr)
library(dplyr) #order matters; load after loading plyr
# for annotation retrieval: 
library(biomaRt)
library(org.Mm.eg.db)
sessionInfo()
#source("https://bioconductor.org/biocLite.R")
#biocLite("org.Mm.eg.db")
#biocLite("BSgenome")
library(BSgenome)
#available.genomes()
#biocLite("BSgenome.Mmusculus.UCSC.mm10.masked")
library(BSgenome.Mmusculus.UCSC.mm10.masked)

# for motif analysis:
library(Biostrings)
library (MotifDb)
require(rtfbs)
```

Read in data from supplement containing micrarray differentially expressed genes.
```{r}
data_DE = read.table("~/Documents/willenbring/motif_analysis_2018_09/Willenbring_Supplementary_Table_DE_MF_vs_Heps.txt", sep="\t", fill=T, header=T)
head(data_DE)

```
Get the chromosome coordinates for each gene's transcription start site (TSS).  
```{r}
ensembl = useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl", host = "useast.ensembl.org")
attributes = listAttributes(ensembl)
#attributes 
annotation_info = getBM(attributes = c("entrezgene","chromosome_name", "transcript_start","transcript_end", "strand"), filters = "entrezgene", values = data_DE$Entrez, mart=ensembl)
head(annotation_info)
```

Find coordinates for the promoters of these genes.
```{r}
#  ------------start-----------------------------------end-->>>>>>>>
# --------transcript_start-100-------------transcript_start->>>>>>>>  (+)

#  <<<<<<<-start-----------------------------------end----------
#  <<<<<<<-transcript_end------------------transcript_end+100---      (-)

annotation_info[, "chrom"] = paste0("chr", annotation_info$chromosome_name)
#replace chromosome MT with M for UCSC naming standards 
annotation_info[which(annotation_info[, "chrom"] == "chrMT"), "chrom"] = "chrM"
annotation_info[, "start"] = unlist(lapply (seq(1, nrow(annotation_info)),  function(x) if (annotation_info[x,"strand"] < 0) { annotation_info[x,"transcript_end"] } else{ annotation_info[x, "transcript_start"] -1000 }  ))
annotation_info[, "end"] = unlist(lapply (seq(1, nrow(annotation_info)),  function(x) if ( annotation_info[x, "strand"] < 0 ) {annotation_info[x, "transcript_end"] +1000} else{annotation_info[x,  "transcript_start"]}  ))
annotation_info[, "strand2"] = unlist(lapply (annotation_info$strand, function(x) if (x < 0) { "-" } else{ "+" } ))

#checked these coordiantes in the genome browser

index = match (data_DE$Entrez, annotation_info$entrezgene)
data_DE2 = cbind(annotation_info[index,], data_DE)

#lots of entrez IDs missing because they look like this "15507 /// 15508"
head(data_DE2)
```


Function to extract relevant promoter sequences for genes of interest

```{r}
get_promoter_sequences <- function (data_DE2){
  myseqs = data_DE2[, c("chrom", "start", "end", "strand2")]
  
  dim(myseqs)
  str(myseqs)
  
  #remove rows with NA
  myseqs = myseqs[complete.cases(myseqs),]
  dim(myseqs)
  
  #remove duplicate rows
  myseqs = distinct(myseqs)
  dim(myseqs)
  
  #remove weird chromosomes 
  myseqs = myseqs[myseqs$chrom %in% paste0("chr", c(seq(1,19), "X", "Y", "MT")),]
  
  myseqs2 = getSeq(BSgenome.Mmusculus.UCSC.mm10.masked, myseqs$chrom, start=myseqs$start, end=myseqs$end, strand=myseqs$strand2)
  
  # give fasta header a more descriptive label
  labels = paste(myseqs$chrom, myseqs$start, myseqs$end, myseqs$strand2, sep="||")
  length(which(!duplicated(labels[order(labels)])))
  names(myseqs2) = labels
  return(myseqs2)
}

# background sequences
myseqs_background = get_promoter_sequences (data_DE2)
writeXStringSet(myseqs_background, "~/Documents/willenbring/background_promoters.fa", format="fasta")
myseqs_background

# foreground sequences 
myseqs_foreground = get_promoter_sequences (data_DE2[which(data_DE2$False.discovery.rate < 0.01),])
writeXStringSet(myseqs_foreground, "~/Documents/willenbring/foreground_promoters.fa", format="fasta")
myseqs_foreground

# background sequences - foreground sequences (used in fisher test below)
myseqs_notforeground = get_promoter_sequences (data_DE2[which(data_DE2$False.discovery.rate >= 0.01),])

```

```{r}
# de novo motif search 

# meme/libexec/meme-5.0.2/fasta-get-markov -dna -m 2 Documents/willenbring/background_promoters.fa Documents/willenbring/background_markov_model.txt
#10187 1001 1001 1001.0 10197187

#meme/bin/meme Documents/willenbring/foreground_promoters.fa -dna -oc ./meme3  -mod zoops -nmotifs 3 -minw 6 -maxw 12 -objfun classic  -bfile Documents/willenbring/background_markov_model.txt

#ZOOPS: Zero or one site per sequence

#convert transfac to meme style format
#meme/libexec/meme-5.0.2/transfac2meme ~/Downloads/transfac_for_tara/match/data/matrix.dat > Downloads/transfac_matrix_dat_to_meme.txt

#TODO
#scan for motifs in jaspar
#motifdiverge
# go back and fix entrez ids that didn't match 

```
Scan for known motifs of genes that are expressed in MFs.  Takes a long time to run. 
```{r}
# redo fisher test (should have compared foreground to background-foreground instead of foreground to background)
expressedgenes = data_DE2[which(!(is.na(data_DE2[,"MFs.1"]))),"Symbol"]  #hopefully these are all genes expressed (assuming someone did the filtering if they are in the excel file)
matches_fg_per_gene = list()
matches_bg_per_gene = list()
tests_per_gene = list()
for ( gene in expressedgenes ) {
  motif = MotifDb [grep (gene, values (MotifDb)$geneSymbol, ignore.case=TRUE)]
  if ( length(motif) >0 ) {
    matches_fg = list()  #multiple motifs per gene 
    matches_bg = list()
    fishertest = list()
    for (x in 1:length(motif)) {
      motif_pwm = motif[[x]]
      matches_fg[[x]] = lapply (myseqs_foreground, function(x) countPWM(motif_pwm, x))
      matches_bg[[x]] = lapply (myseqs_notforeground, function(x) countPWM(motif_pwm, x))
      test_results = fisher.test( cbind(c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0)) )) )
      fishertest[[x]] = c(gene, x, test_results$p.value, c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0))) )
    }
    matches_fg_per_gene[[gene]] = matches_fg
    matches_bg_per_gene[[gene]] = matches_bg
    tests_per_gene[[gene]] = fishertest
  }
}
all_fisher_tests = matrix( unlist(tests_per_gene), nc=7, byrow =T)
colnames(all_fisher_tests) = c("TF", "TF_motif_num", "fisher_test_pvalue", "fg_hit", "fg_no_hit", "bg_hit", "bg_no_hit")
```

Save results 
```{r}
#save(matches_fg_per_gene, matches_bg_per_gene, tests_per_gene, file = "Documents/Willenbring/fisher_second_attempt.Robj")
#write.table(all_fisher_tests, "Documents/Willenbring/motif_analysis_2018_09/TF_motif_fisher_tests_v2.xls", sep="\t", quote=F, row.names = F, col.names = c("TF", "TF_motif_num", "fisher_test_pvalue", "fg_hit", "fg_no_hit", "bg_hit", "bg_no_hit"))
```

We can upload the results of the fisher test instead of re-running it.  
```{r}
load("~/Documents/Willenbring/motif_analysis_2018_09/fisher_second_attempt.Robj")

#table of motif counts per promoter/gene
temp = do.call(paste, c(matches_fg_per_gene, names(matches_fg_per_gene)))

```

```{r}
#investigate this. probably because not all genes could be found in ensembl
length(myseqs_foreground) + length(myseqs_notforeground) != length(myseqs_background)

```
Adjust p-values. 
```{r}
fisher_test_multiple_testing_correction = read.table("~/Documents/Willenbring/motif_analysis_2018_09/all_fisher_tests_v2.xls", header=T)
fisher_test_multiple_testing_correction[,'corrected_pvalue'] = p.adjust(fisher_test_multiple_testing_correction$fisher_test_pvalue, method="BH")

length(which(fisher_test_multiple_testing_correction$corrected_pvalue <= 0.05))

write.table(fisher_test_multiple_testing_correction, "Documents/Willenbring/motif_analysis_2018_09/TF_motif_fisher_tests_v2_adjusted.xls", sep="\t", quote=F, row.names = F, col.names = T)
}   
```


```{r}

load("~/Documents/willenbring/motif_analysis_2018_09/fisher_second_attempt.Robj")
unzip_tf_locations <- function (matches____per_gene){
  tf_matrix = list()
  inc = 1
  for (tf in matches____per_gene){
    tf_matrix[[inc]] =  matrix(unlist(tf), byrow=T, nc=length(tf))
    inc = inc + 1
  }

  tf_names = list()
  inc = 1
  for (tfname in  names(matches____per_gene)){
    length_value = length(matches____per_gene[[tfname]])
    for (x in 1:length_value){
      tf_names[[inc]] = paste(tfname,x, sep="_" )
      inc = inc + 1
    }
  }
  
  tf_matrix2 = do.call(cbind, tf_matrix)
  dim(tf_matrix2)
  rownames(tf_matrix2) = names(matches____per_gene$Ace[[1]])
  colnames(tf_matrix2) = unlist(tf_names)
  
  return(tf_matrix2)  # motif count per location 
}

tf_matrix2_fg = unzip_tf_locations(matches_fg_per_gene)
tf_matrix2_bg = unzip_tf_locations(matches_bg_per_gene)
```

```{r}
genes = data_DE2

#remove rows where I couldn't link a gene name to a location
genes2 = genes[which(!is.na(genes$chromosome_name)),]

tf_matrix2 = rbind(tf_matrix2_fg, tf_matrix2_bg)
promoter_names = paste(genes2$chrom, genes2$start, genes2$end, genes2$strand2, sep="||")
index = match( rownames(tf_matrix2), promoter_names)
#cbind(rownames(tf_matrix2), promoter_names[index])

#y values
y_regression = genes2[index, "Fold.change..log2."]


y_classifier = rep(-1, length(y_regression))
y_classifier [ which(genes2[index, "False.discovery.rate"] <0.05)] <- 1
mean = mean(genes2[index, "False.discovery.rate"],na.rm=T)
y_classifier [which(genes2[index, "False.discovery.rate"] <mean & genes2[index, "False.discovery.rate"] >=0.05)] <- NA
table(y_classifier)
y_classifier2 = y_classifier[which(!is.na(y_classifier) & !is.na(y_regression))]
table(y_classifier2)

#remove NA
y_regression2 = y_regression[which(!is.na(y_classifier) & !is.na(y_regression))]

#remove NA labels
tf_matrix3 = tf_matrix2[which(!is.na(y_classifier) & !is.na(y_regression)),]
tf_matrix4 = as.matrix(tf_matrix3)

#check dimensions to see if they match 
dim(tf_matrix4)
length(y_regression2)
length(y_classifier2)

```

```{r}
require(randomForest)
train=sample(1:nrow(tf_matrix4),300)
exp.rf=randomForest(tf_matrix4, y2, subset = train)
save(exp.rf, file = "Documents/willenbring/first_random_forest.Robj")
```

```{r}

```

```{r}
#write to file to run in python 
write.table(tf_matrix4, "~/Downloads/X_values.xls", sep="\t", col.names=T, row.names=F)
write.table(y_regression2, "~/Downloads/y_values.xls", sep="\t", col.names=T, row.names=F)
write.table(y_classifier2, "~/Downloads/y_values_binary.xls", sep="\t", col.names=T, row.names=F)

```
#error goes down when we add more trees
```{r}
exp.rf
plot(exp.rf)
importanceOrder=order(-exp.rf$importance)
names=rownames(exp.rf$importance)[importanceOrder][1:50]
names
varImpPlot(exp.rf)
testset = tf_matrix4[which(!(1:nrow(tf_matrix4) %in% train)),]
Prediction <- predict(exp.rf, testset)
cor(Prediction, y2[which(!(1:nrow(tf_matrix4) %in% train))])
table(observed=y2[which(!(1:nrow(tf_matrix4) %in% train))],predicted=Prediction)


par(mfrow=c(5, 3), xpd=NA)
for (name in names)
+   partialPlot(exp.rf, df, eval(name), main=name, xlab=name,ylim=c(-.2,.9))
```
4800

```{r}
temp = read.table("~/Downloads/output_regression.txt")
colnames(tf_matrix4)[which (temp$V1>0)]
barplot(temp[which(temp$V1>0.003), "V1"], names.arg = colnames(tf_matrix4)[which (temp$V1>0.003)], las=2   )
temp = read.table("~/Downloads/output_classifier.txt")
colnames(tf_matrix4)[which (temp$V1>0)]

barplot(temp[which(temp$V1>0.003), "V1"], names.arg = colnames(tf_matrix4)[which (temp$V1>0.003)], las=2   )
rbind (colnames(tf_matrix4)[which (temp$V1>0)],temp[which(temp$V1>0.0), "V1"] )
```
```{r}
expressedgenes = data_DE2[which(!(is.na(data_DE2[,"MFs.1"]))),"Symbol"]  
for ( gene in expressedgenes ) {
  motif = MotifDb [grep (gene, values (MotifDb)$geneSymbol, ignore.case=TRUE)]
  if ( length(motif) >0 ) {
    for (x in 1:length(motif)) {
      pdf(paste0("~/Downloads/motifs/motif_", gene, "_", str(x), ".pdf"))
      seqLogo (motif[[x]])
      dev.off()
    }
  }
}
```
