---
title: "2018_10_22_Willenbring_motif"
output: html_document
---
```{r }
library(plyr)
library(dplyr) #order matters; load after loading plyr
# for annotation retrieval: 
library(biomaRt)
library(org.Mm.eg.db)
sessionInfo()
#source("https://bioconductor.org/biocLite.R")
#biocLite("org.Mm.eg.db")
#biocLite("BSgenome")
library(BSgenome)
#available.genomes()
#biocLite("BSgenome.Mmusculus.UCSC.mm10.masked")
library(BSgenome.Mmusculus.UCSC.mm10.masked)

# for motif analysis:
library(Biostrings)
library (MotifDb)
require(rtfbs)
```

```{r}
data_DE = read.delim("~/Documents/willenbring/motif_analysis_2018_09/Willenbring_Supplementary_Table_DE_MF_vs_Heps.txt", sep="\t", fill=T, header=T)
colnames(data_DE)
dim(data_DE)
data_DE$Entrez2 = strsplit2(as.character(data_DE$Entrez), " /// ")[,1]
```
Looking for genes that are different betwee MF and Heps & iHeps and between Heps and iHeps. 
```{r}
length(which(data_DE[, "False.discovery.rate"]<0.05))
length(which(data_DE[, "False.discovery.rate.1"]<0.05))
length(which(data_DE[, "False.discovery.rate.2"]<0.20))
index1 = data_DE[, "False.discovery.rate"]<0.05
index2 = data_DE[, "False.discovery.rate.1"]<0.05
index3 = data_DE[, "False.discovery.rate.2"]<0.20
index4 = complete.cases(data_DE[,c( "MFs.1", "MFs.2", "MFs.3", "Heps1", "Heps2", "Heps3", "MF.iHeps1", "MF.iHeps2", "MF.iHeps3", "MF.iHeps4", "MF.iHeps5")])

focus = data_DE[which(index1 & index2 & index3 & index4), c( "MFs.1", "MFs.2", "MFs.3", "Heps1", "Heps2", "Heps3", "MF.iHeps1", "MF.iHeps2", "MF.iHeps3", "MF.iHeps4", "MF.iHeps5")]
                 
focus = data.matrix(focus)
dim(focus)

pheatmap(focus, labels_row=data_DE[which(index1 & index2 & index3 & index4),"Symbol"])     

```
Get the chromosome coordinates for each gene's transcription start site (TSS).  
```{r}
ensembl = useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl", host = "useast.ensembl.org")
attributes = listAttributes(ensembl)
#attributes 
annotation_info = getBM(attributes = c("entrezgene","chromosome_name", "transcript_start","transcript_end", "strand"), filters = "entrezgene", values = as.character(data_DE$Entrez2), mart=ensembl)
head(annotation_info)
```


```{r}
#  ------------start-----------------------------------end-->>>>>>>>
# --------transcript_start-100-------------transcript_start->>>>>>>>  (+)

#  <<<<<<<-start-----------------------------------end----------
#  <<<<<<<-transcript_end------------------transcript_end+100---      (-)

annotation_info[, "chrom"] = paste0("chr", annotation_info$chromosome_name)
#replace chromosome MT with M for UCSC naming standards 
annotation_info[which(annotation_info[, "chrom"] == "chrMT"), "chrom"] = "chrM"
annotation_info[, "start"] = unlist(lapply (seq(1, nrow(annotation_info)),  function(x) if (annotation_info[x,"strand"] < 0) { annotation_info[x,"transcript_end"] } else{ annotation_info[x, "transcript_start"] -1000 }  ))
annotation_info[, "end"] = unlist(lapply (seq(1, nrow(annotation_info)),  function(x) if ( annotation_info[x, "strand"] < 0 ) {annotation_info[x, "transcript_end"] +1000} else{annotation_info[x,  "transcript_start"]}  ))
annotation_info[, "strand2"] = unlist(lapply (annotation_info$strand, function(x) if (x < 0) { "-" } else{ "+" } ))

#checked these coordiantes in the genome browser

index = match (data_DE$Entrez2, annotation_info$entrezgene)
data_DE2 = cbind(annotation_info[index,], data_DE)

#lots of entrez IDs missing because they look like this "15507 /// 15508"
# picking the first of the two for now. 
#probe probably binds to both but it I am not sure I want to include both or just pick one. 
head(data_DE2)
```


```{r}
get_promoter_sequences <- function (data_DE2){
  myseqs = data_DE2[, c("chrom", "start", "end", "strand2")]
  
  dim(myseqs)
  str(myseqs)
  
  #remove rows with NA
  myseqs = myseqs[complete.cases(myseqs),]
  dim(myseqs)
  
  #remove duplicate rows
  myseqs = distinct(myseqs)
  dim(myseqs)
  
  #remove weird chromosomes 
  myseqs = myseqs[myseqs$chrom %in% paste0("chr", c(seq(1,19), "X", "Y", "MT")),]
  
  myseqs2 = getSeq(BSgenome.Mmusculus.UCSC.mm10.masked, myseqs$chrom, start=myseqs$start, end=myseqs$end, strand=myseqs$strand2)
  
  # give fasta header a more descriptive label
  labels = paste(myseqs$chrom, myseqs$start, myseqs$end, myseqs$strand2, sep="||")
  length(which(!duplicated(labels[order(labels)])))
  names(myseqs2) = labels
  return(myseqs2)
}

# background sequences
myseqs_background = get_promoter_sequences (data_DE2)
writeXStringSet(myseqs_background, "~/Documents/willenbring/background_promoters_version2.fa", format="fasta")
myseqs_background

# foreground sequences 
myseqs_foreground = get_promoter_sequences (data_DE2[which(index1 & index2 & index3),])
writeXStringSet(myseqs_foreground, "~/Documents/willenbring/foreground_promoters_version2.fa", format="fasta")
myseqs_foreground

```
meme/libexec/meme-5.0.2/fasta-get-markov -dna -m 2 Documents/willenbring/background_promoters.fa Documents/willenbring/background_markov_model.txt
19541 1001 1001 1001.0 19560541
DOM-4DHJGH6-LT:~ tfriedrich$ 
Scan for known motifs of genes that are expressed in MFs.  Takes a long time to run. 
```{r}
# redo fisher test (should have compared foreground to background-foreground instead of foreground to background)
expressedgenes = data_DE2[which(!(is.na(data_DE2[,"MFs.1"]))),"Symbol"]  #hopefully these are all genes expressed (assuming someone did the filtering if they are in the excel file)
matches_fg_per_gene = list()
matches_bg_per_gene = list()
tests_per_gene = list()
for ( gene in expressedgenes ) {
  motif = MotifDb [grep (gene, values (MotifDb)$geneSymbol, ignore.case=TRUE)]
  if ( length(motif) >0 ) {
    matches_fg = list()  #multiple motifs per gene 
    matches_bg = list()
    fishertest = list()
    for (x in 1:length(motif)) {
      motif_pwm = motif[[x]]
      matches_fg[[x]] = lapply (myseqs_foreground, function(x) countPWM(motif_pwm, x))
      matches_bg[[x]] = lapply (myseqs_notforeground, function(x) countPWM(motif_pwm, x))
      test_results = fisher.test( cbind(c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0)) )) )
      fishertest[[x]] = c(gene, x, test_results$p.value, c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0))) )
    }
    matches_fg_per_gene[[gene]] = matches_fg
    matches_bg_per_gene[[gene]] = matches_bg
    tests_per_gene[[gene]] = fishertest
  }
}
all_fisher_tests = matrix( unlist(tests_per_gene), nc=7, byrow =T)
colnames(all_fisher_tests) = c("TF", "TF_motif_num", "fisher_test_pvalue", "fg_hit", "fg_no_hit", "bg_hit", "bg_no_hit")
```

Adjust p-values. 
```{r}

all_fisher_tests_corr_pvalue= p.adjust(as.numeric(all_fisher_tests[,"fisher_test_pvalue"]), method="fdr")

length(which(all_fisher_tests_corr_pvalue <= 0.05))

write.table(cbind(all_fisher_tests, all_fisher_tests_corr_pvalue), "/Users/tfriedrich/Documents/Willenbring/TF_motif_fisher_tests_v3_adjusted.xls", sep="\t", quote=F, row.names = F, col.names = T)
  
```

