---
title: "2018_09_07_Willenbring_motif_enrichment"
output: html_document
---

setup packages
```{r }
library(plyr)
library(dplyr) #order matters; load after loading plyr
library(biomaRt)
library(org.Mm.eg.db)
sessionInfo()
#source("https://bioconductor.org/biocLite.R")
#biocLite("org.Mm.eg.db")
#biocLite("BSgenome")
library(BSgenome)
#available.genomes()
#biocLite("BSgenome.Mmusculus.UCSC.mm10.masked")
library(BSgenome.Mmusculus.UCSC.mm10.masked)
```

Read in data from supplement containing micrarray differentially expressed genes.
```{r}
data_DE = read.table("~/Documents/willenbring/motif_analysis_2018_09/Willenbring_Supplementary_Table_DE_MF_vs_Heps.txt", sep="\t", fill=T, header=T)
head(data_DE)

```
Get the chromosome coordinates for each gene's transcription start site (TSS).  
```{r}
ensembl = useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl", host = "useast.ensembl.org")
attributes = listAttributes(ensembl)
#attributes 
annotation_info = getBM(attributes = c("entrezgene","chromosome_name", "transcript_start","transcript_end", "strand"), filters = "entrezgene", values = data_DE$Entrez, mart=ensembl)
head(annotation_info)
```

Find coordinates for the promoters of these genes.
```{r}
#  ------------start-----------------------------------end-->>>>>>>>
# --------transcript_start-100-------------transcript_start->>>>>>>>  (+)

#  <<<<<<<-start-----------------------------------end----------
#  <<<<<<<-transcript_end------------------transcript_end+100---      (-)

annotation_info[, "chrom"] = paste0("chr", annotation_info$chromosome_name)
#replace chromosome MT with M for UCSC naming standards 
annotation_info[which(annotation_info[, "chrom"] == "chrMT"), "chrom"] = "chrM"
annotation_info[, "start"] = unlist(lapply (seq(1, nrow(annotation_info)),  function(x) if (annotation_info[x,"strand"] < 0) { annotation_info[x,"transcript_end"] } else{ annotation_info[x, "transcript_start"] -1000 }  ))
annotation_info[, "end"] = unlist(lapply (seq(1, nrow(annotation_info)),  function(x) if ( annotation_info[x, "strand"] < 0 ) {annotation_info[x, "transcript_end"] +1000} else{annotation_info[x,  "transcript_start"]}  ))
annotation_info[, "strand2"] = unlist(lapply (annotation_info$strand, function(x) if (x < 0) { "-" } else{ "+" } ))

index = match (data_DE$Entrez, annotation_info$entrezgene)
data_DE2 = cbind(annotation_info[index,], data_DE)

#lots of entrez IDs missing because they look like this "15507 /// 15508"
head(data_DE2)
```




```{r}
### Bimap interface:
#x <- org.Mm.egCHRLOC
## Get the entrez gene identifiers that are mapped to chromosome locations
#mapped_genes <- mappedkeys(x)
## Convert to a list
#xx <- as.list(x[mapped_genes])
#if(length(xx) > 0) {
#      # Get the CHRLOC for the first five genes
#      xx[1:5]
#      # Get the first one
#      xx[[1]]
#}

# x_END <- org.Mm.egCHRLOCEND
# mapped_genes_END <- mappedkeys(x_END)
# xx_END <- as.list(x[mapped_genes_END])
# 
# annotation_END = ldply(xx_END, data.frame)
# annotation = ldply(xx, data.frame)
# chr = unlist(lapply(xx, names))
# annotation2 = cbind(annotation, annotation_END, chr)
# colnames(annotation2) = c("gene_id1", "start_loc", "gene_id2", "end_loc", "chr")


#  --------------------------X-->>>>>>>>
#  ----start----------------end->>>>>>>>  (+)

#  <<<<<<<-X------------------------
#  <<<<<<<-start----------------end-      (-)

# annotation2[, "chrom"] = paste0("chr", annotation2$chr)
# annotation2[, "start"] = unlist(lapply (annotation2$start_loc, function(x) abs(x) - 1000  ))
# annotation2[, "end"] = unlist(lapply (annotation2$start_loc, function(x) abs(x) ))
# annotation2[, "strand"] = unlist(lapply (annotation2$start_loc, function(x) if (x < 0) { "-" } else{ "+" } ))
# 
# index = match (data_DE$Entrez, annotation2$.id)
# data_DE2 = cbind(annotation2[index,], data_DE)
# 
# #lots of entrez IDs missing because they look like this "15507 /// 15508"
# head(data_DE2)
```

Function to extract relevant promoter sequences for genes of interest

```{r}
get_promoter_sequences <- function (data_DE2){
  myseqs = data_DE2[, c("chrom", "start", "end", "strand2")]
  
  dim(myseqs)
  str(myseqs)
  
  #remove rows with NA
  myseqs = myseqs[complete.cases(myseqs),]
  dim(myseqs)
  
  #remove duplicate rows
  myseqs = distinct(myseqs)
  dim(myseqs)
  
  #remove weird chromosomes 
  myseqs = myseqs[myseqs$chrom %in% paste0("chr", c(seq(1,19), "X", "Y", "MT")),]
  
  myseqs2 = getSeq(BSgenome.Mmusculus.UCSC.mm10.masked, myseqs$chrom, start=myseqs$start, end=myseqs$end, strand=myseqs$strand2)
  
  # give fasta header a more descriptive label
  labels = paste(myseqs$chrom, myseqs$start, myseqs$end, myseqs$strand2, sep="||")
  length(which(!duplicated(labels[order(labels)])))
  names(myseqs2) = labels
  return(myseqs2)
}

# background sequences
myseqs_background = get_promoter_sequences (data_DE2)
writeXStringSet(myseqs_background, "~/Documents/willenbring/background_promoters.fa", format="fasta")
myseqs_background

# foreground sequences 
myseqs_foreground = get_promoter_sequences (data_DE2[which(data_DE2$False.discovery.rate < 0.01),])
writeXStringSet(myseqs_foreground, "~/Documents/willenbring/foreground_promoters.fa", format="fasta")
myseqs_foreground

# background sequences - foreground sequences (used in fisher test below)
myseqs_notforeground = get_promoter_sequences (data_DE2[which(data_DE2$False.discovery.rate >= 0.01),])

```

```{r}

# de novo motif search 
#meme foreground_promoters.fa -dna -oc . -nostatus -time 18000 -mod zoops -nmotifs 3 -minw 6 -maxw 20 -objfun de -neg background_promoters.fa -revcomp -markov_order 0
#meme ran successfully in 2393.78 seconds
#####memesuitetool
######meme fasta.fa -dna -oc . -nostatus -time 18000 -mod zoops -nmotifs 3 -minw 6 -maxw 14 -objfun classic -revcomp -markov_order 
###################

# meme/libexec/meme-5.0.2/fasta-get-markov -dna -m 2 Documents/willenbring/background_promoters.fa Documents/willenbring/background_markov_model.txt
#10187 1001 1001 1001.0 10197187

#meme/bin/meme Documents/willenbring/foreground_promoters.fa -dna -oc ./meme3  -mod zoops -nmotifs 3 -minw 6 -maxw 12 -objfun classic  -bfile Documents/willenbring/background_markov_model.txt

#convert transfac to meme style format
#meme/libexec/meme-5.0.2/transfac2meme ~/Downloads/transfac_for_tara/match/data/matrix.dat > Downloads/transfac_matrix_dat_to_meme.txt
#ZOOPS: Zero or one site per sequence

#scan for motifs in jaspar

#motifdiverge

#motif enrichment using meme 




#find more tf that are expressed

# go back and fix entrez ids that didn't match 


```

```{r}
#scan for motifs in transfac 
#Foxa3, Gata4, and Hnf1a or Foxa1/Foxa2/Foxa3 and Hnf4a 

library(Biostrings)
library (MotifDb)
require(rtfbs)

#pwm <- read.pwm("~/Documents/common_resources/transfac_matrix_dat_to_meme.txt")
#set = query (MotifDb, 'hsapiens') 
set = query (MotifDb, 'mmusculus') 
set2 = query(set, "FoxA3")
set3 = query(set, "HNF1A")
set4 = query(set, "GATA4")

fg2 = lapply (myseqs_foreground, function(x) matchPWM(set2[[1]], x))
bg2 = lapply (myseqs_background, function(x) matchPWM(set2[[1]], x))

fg3 = lapply (myseqs_foreground, function(x) matchPWM(set3[[1]], x))
bg3 = lapply (myseqs_background, function(x) matchPWM(set3[[1]], x))

fg4 = lapply (myseqs_foreground, function(x) matchPWM(set4[[1]], x))
bg4 = lapply (myseqs_background, function(x) matchPWM(set4[[1]], x))

fg2n = lapply (myseqs_foreground, function(x) countPWM(set2[[1]], x))
bg2n = lapply (myseqs_background, function(x) countPWM(set2[[1]], x))

fg3n = lapply (myseqs_foreground, function(x) countPWM(set3[[1]], x))
bg3n = lapply (myseqs_background, function(x) countPWM(set3[[1]], x))

fg4n = lapply (myseqs_foreground, function(x) countPWM(set4[[1]], x))
bg4n = lapply (myseqs_background, function(x) countPWM(set4[[1]], x))

```

```{r}
sum(unlist(fg4n))
sum(unlist(fg3n))
sum(unlist(fg2n))

sum(unlist(bg4n))
sum(unlist(bg3n))
sum(unlist(bg2n))
       
length(myseqs_background)
length(myseqs_foreground)


fisher.test( cbind(c(length(which(fg4n>0)), length(which(fg4n==0))), c(length(which(bg4n>0)), length(which(bg4n==0)) )))

fisher.test( cbind(c(length(which(fg2n>0)), length(which(fg2n==0))), c(length(which(bg2n>0)), length(which(bg2n==0)) )) )

```
```{r}
# 
# expressedgenes = data_DE2[which(!(is.na(data_DE2[,"MFs.1"]))),"Symbol"]  #hopefully these are all genes expressed (assuming someone did the filtering if they are in the excel file)
# matches_fg_per_gene = list()
# matches_bg_per_gene = list()
# tests_per_gene = list()
# for ( gene in expressedgenes ) {
#   motif = MotifDb [grep (gene, values (MotifDb)$geneSymbol, ignore.case=TRUE)]
#   if ( length(motif) >0 ) {
#     matches_fg = list()  #multiple motifs per gene 
#     matches_bg = list()
#     test = list()
#     for (x in 1:length(motif)) {
#       motif_pwm = motif[[x]]
#       matches_fg[[x]] = lapply (myseqs_foreground, function(x) countPWM(motif_pwm, x))
#       matches_bg[[x]] = lapply (myseqs_background, function(x) countPWM(motif_pwm, x))
#       test_results = fisher.test( cbind(c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0)) )) )
#       test[[x]] = c(gene, x, test_results$p.value, c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0))) )
#     }
#     matches_fg_per_gene[[gene]] = matches_fg
#     matches_bg_per_gene[[gene]] = matches_bg
#     tests_per_gene[[gene]] = test
#     
#     
#   }
# }
# 
# all_fisher_tests = matrix( unlist(tests_per_gene), nc=7, byrow =T)
# # 
```

```{r}
#write.table(all_fisher_tests, "Documents/Willenbring/motif_analysis_2018_09/TF_motifs_fisher_tests.xls", sep="\t", quote=F, row.names = F, col.names = c("TF", "TF_motif_num", "fisher_test_pvalue", "fg_hit", "fg_no_hit", "bg_hit", "bg_no_hit"))
#matches_fg_per_gene$Hnf4a[[1]]

#save(matches_fg_per_gene, matches_bg_per_gene, tests_per_gene, file = "Documents/Willenbring/motif_analysis_2018_09/fisher_first_attempt.Robj")
```

Scan for known motifs of genes that are expressed in MFs.  Takes a long time to run. 
```{r}
# redo fisher test (should have compared foreground to background-foreground instead of foreground to background)
expressedgenes = data_DE2[which(!(is.na(data_DE2[,"MFs.1"]))),"Symbol"]  #hopefully these are all genes expressed (assuming someone did the filtering if they are in the excel file)
matches_fg_per_gene = list()
matches_bg_per_gene = list()
tests_per_gene = list()
for ( gene in expressedgenes ) {
  motif = MotifDb [grep (gene, values (MotifDb)$geneSymbol, ignore.case=TRUE)]
  if ( length(motif) >0 ) {
    matches_fg = list()  #multiple motifs per gene 
    matches_bg = list()
    fishertest = list()
    for (x in 1:length(motif)) {
      motif_pwm = motif[[x]]
      matches_fg[[x]] = lapply (myseqs_foreground, function(x) countPWM(motif_pwm, x))
      matches_bg[[x]] = lapply (myseqs_notforeground, function(x) countPWM(motif_pwm, x))
      test_results = fisher.test( cbind(c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0)) )) )
      fishertest[[x]] = c(gene, x, test_results$p.value, c(length(which(matches_fg[[x]]>0)), length(which(matches_fg[[x]]==0))), c(length(which(matches_bg[[x]]>0)), length(which(matches_bg[[x]]==0))) )
    }
    matches_fg_per_gene[[gene]] = matches_fg
    matches_bg_per_gene[[gene]] = matches_bg
    tests_per_gene[[gene]] = fishertest
  }
}
all_fisher_tests = matrix( unlist(tests_per_gene), nc=7, byrow =T)
colnames(all_fisher_tests) = c("TF", "TF_motif_num", "fisher_test_pvalue", "fg_hit", "fg_no_hit", "bg_hit", "bg_no_hit")
```
Save results 
```{r}
#save(matches_fg_per_gene, matches_bg_per_gene, tests_per_gene, file = "Documents/Willenbring/fisher_second_attempt.Robj")
#write.table(all_fisher_tests, "Documents/Willenbring/all_fisher_tests_v2.xls", sep="\t", quote=F, row.names = F, col.names = c("TF", "TF_motif_num", "fisher_test_pvalue", "fg_hit", "fg_no_hit", "bg_hit", "bg_no_hit"))
```

```{r}
load("~/Documents/Willenbring/motif_analysis_2018_09/fisher_second_attempt.Robj")

#table of motif counts per promoter/gene
temp = do.call(paste, c(matches_fg_per_gene, names(matches_fg_per_gene)))

```
Upload the results 
```{r}
#investigate this. probably because not all genes could be found in ensembl
length(myseqs_foreground) + length(myseqs_notforeground) != length(myseqs_background)

#all_fisher_tests[,"fg_hit"] = as.numeric(as.character(all_fisher_tests[,"fg_hit"]))
#all_fisher_tests[,"fg_no_hit"] = as.numeric(as.character(all_fisher_tests[,"fg_no_hit"]))
#all_fisher_tests[,"bg_hit"] = as.numeric(as.character(all_fisher_tests[,"bg_hit"]))
#all_fisher_tests[,"bg_no_hit"] = as.numeric(as.character(all_fisher_tests[,"bg_no_hit"]))

#fisher = lapply(seq(1, nrow(all_fisher_tests)), function(x)  cbind (c(all_fisher_tests[x, "fg_hit"], all_fisher_tests[x, "fg_no_hit"]), c(all_fisher_tests[x, "bg_hit"], all_fisher_tests[x, "bg_no_hit"]) ))
```

