---
title: "FC_reanalysis with markers of interest"
author: "Tara Friedrich"
date: "6/10/2019"
output: html_document
---

```{r setup, include=FALSE}
library(Seurat)
library(garnett)
library(monocle)
library(org.Mm.eg.db)
sessionInfo()
```
Read in data and filter based on minimum number of cells and genes. 
```{r}
cellsdata = Read10X("/Users/tfriedrich/Downloads/FC_102418_with_markers/outs/filtered_feature_bc_matrix/")
cells = CreateSeuratObject(counts = cellsdata, project = "FC", min.cells = 3, min.features = 200)
```
Number of each cell showing expression for each marker. 
```{r}
length(which(cellsdata["EGFP.dna",]>0))
length(which(cellsdata["lacZ.dna",]>0))
length(which(cellsdata["mCerulean.dna",]>0))
length(which(cellsdata["mCherry.dna",]>0))
length(which(cellsdata["mOrange.dna",]>0))
length(which(cellsdata["tdrfp.dna",]>0))
```
Write cells expressing markers of interest to a file. 
```{r}
A = subset(subset(cells, lacZ.dna>0) ,Xist>0)
B = subset(subset(cells, lacZ.dna==0) ,Xist>0)
C = subset(subset(cells, tdrfp.dna>0) ,Xist=0)
tdrfp = subset(cells, tdrfp.dna>0)
lacZ = subset(cells, lacZ.dna>0)
mOrange = subset(cells, mOrange.dna>0)
mCerulean = subset(cells, mCerulean.dna>0)
mCherry = subset(cells, mCherry.dna>0)
EGFP = subset(cells, EGFP.dna>0)

write.csv(x = GetAssayData(A), file = "lacZ_positive_and_Xist_positive.csv", quote=FALSE)
write.csv(x = GetAssayData(B), file = "lacZ_negative_and_Xist_positive.csv", quote=FALSE)
write.csv(x = GetAssayData(C), file = "tdrfp_positve_and_Xist_negative.csv", quote=FALSE)

write.csv(x = GetAssayData(tdrfp), file = "tdrfp_positive.csv", quote=FALSE)
write.csv(x = GetAssayData(lacZ), file = "lacZ_positive.csv", quote=FALSE)
write.csv(x = GetAssayData(mOrange), file = "mOrange_positve.csv", quote=FALSE)
write.csv(x = GetAssayData(mCerulean), file = "mCerulean_positve.csv", quote=FALSE)
write.csv(x = GetAssayData(mCherry), file = "mCherry_positve.csv", quote=FALSE)
write.csv(x = GetAssayData(EGFP), file = "EGFP_positve.csv", quote=FALSE)
```
Gather scRNA-seq QC metric.
```{r}
mito.genes <- grep(pattern = "^mt-", x = rownames(x = cells), value = TRUE)
percent.mito <- Matrix::colSums(cells[mito.genes, ])/Matrix::colSums(cells)
```

AddMetaData adds columns to object@meta.data, and is a great place to stash QC stats
```{r}
cells <- AddMetaData(object = cells, metadata = percent.mito, col.name = "percent.mito")
```

Plot expressino of marker genes
```{r}
VlnPlot(object = cells, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"))
marker_genes = c("Hnf4a", "Alb", "Fah", "Krt8", "Krt18", "G6pc", "Sox9", "Krt7", "Krt19", "Epcam", "Onecut1", "Afp", "Xist")
VlnPlot(A,c("Hnf4a", "Alb", "Fah"))
VlnPlot(B,c("Hnf4a", "Alb", "Fah"))
VlnPlot(C,c("Hnf4a", "Alb", "Fah"))
```

```{r}
cells <- NormalizeData(object = cells, normalization.method = "LogNormalize", 
    scale.factor = 10000)
```
vst: First, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression (loess). Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). Feature variance is then calculated on the standardized values after clipping to a maximum (see clip.max parameter).
```{r}
cells <- FindVariableFeatures(object = cells, selection.method = "vst", nfeatures = 2000)
```

Regress out expression effects confounded by nUMI and percent.mito
```{r}
cells <- ScaleData(object = cells, vars.to.regress = c("nUMI", "percent.mito"))
```
Regress out cell cycle genes.
For each gene, Seurat models the relationship between gene expression and the S and G2M cell cycle scores. The scaled residuals of this model represent a ‘corrected’ expression matrix, that can be used downstream for dimensional reduction.

```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

cells <- CellCycleScoring(object=cells, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
```

```{r}
cells <- RunPCA(object = cells, pc.genes = cells@var.genes, do.print = TRUE, pcs.print = 1:5, genes.print = 5)
ElbowPlot(cells)
```

Highly variable genes. 
```{r}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(cells), 10)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(cells)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2

```
Visualize using UMAP. Colored by cell state. 
```{r}
cells <- RunUMAP(object = cells, dims = 1:10)
```

```{r}
#pdf("FeaturePlots_FC.pdf")
DimPlot(object = cells, reduction = "umap")
FeaturePlot(cells, features = marker_genes[1:4])
FeaturePlot(cells, features = marker_genes[5:8])
FeaturePlot(cells, features = marker_genes[9:12])
FeaturePlot(cells, features = marker_genes[13])
FeaturePlot(cells, features = c("EGFP.dna","lacZ.dna","mCerulean.dna","mCherry.dna"))
FeaturePlot(cells, features = c("mCherry.dna","mOrange.dna","tdrfp.dna"))
#dev.off()
```

```{r}
#Extract data, phenotype data, and feature data from the SeuratObject
#https://github.com/cole-trapnell-lab/monocle-release/issues/262
data <- as(as.matrix(cells@assays$RNA@data), 'sparseMatrix')
pd <- new('AnnotatedDataFrame', data = cells@meta.data)
fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
fd <- new('AnnotatedDataFrame', data = fData)
monocle_cds <- newCellDataSet(data,
                         phenoData = pd,
                         featureData = fd,
                         lowerDetectionLimit = 0.5,
                         expressionFamily = negbinomial.size())

# generate size factors for normalization later
monocle_cds <- estimateSizeFactors(monocle_cds)

# markers for common cell types
marker_file_path <-  "/Users/tfriedrich/Downloads/markers_garnett.txt"
marker_check <- check_markers(monocle_cds, marker_file_path,
                              db=org.Mm.eg.db,
                              cds_gene_id_type = "SYMBOL",
                              marker_file_gene_id_type = "SYMBOL")
plot_markers(marker_check)
```
# train cells using marker genes as labels
Garnett classification is trained using a multinomial elastic-net regression. This means that certain genes are chosen as the relevant genes for distinguishing between cell types. Which genes are chosen may be of interest, so Garnett includes a function to access the chosen genes. Note: Garnett does not regularize the input markers, so they will be included in the classifier regardless.
```{r}
set.seed(260)
pbmc_classifier <- train_cell_classifier(cds = monocle_cds,
                                         marker_file = marker_file_path,
                                         db=org.Mm.eg.db,
                                         cds_gene_id_type = "SYMBOL",
                                         num_unknown = 50,
                                         marker_file_gene_id_type = "SYMBOL")
head(pData(monocle_cds))
feature_genes <- get_feature_genes(pbmc_classifier,
                                   node = "root",
                                   db = org.Mm.eg.db, 
                                   convert_ids = TRUE)
head(feature_genes)

write.table(feature_genes, "/Users/tfriedrich/Downloads/garnett_elastic_regression_feature_genes.xls", sep="\t", col.names=NA)
```

Differential Expression between TRFP positive and egfp/mCerulean/mCherry/mOrange positive cells 
```{r}

library(edgeR)
DE <- FindMarkers(cells, ident.1 = "G1", ident.2 = "S")

MarkerA  = GetAssayData(EGFP) 
MarkerB  = GetAssayData(tdrfp)
  
```
