---
output:
  html_document: default
  pdf_document: default
---
---
title: "Chen lab RNA-seq version 3 (S449 Samples redone by Novogene again) "
output: html_notebook
date: April 5, 2019
#human cell lines 
# si RNA knockdown experiment
```{r}
library(pheatmap)
library(edgeR)
library(DESeq2)
library(biomaRt)
library(ggplot2)
sessionInfo()
```
Read in files. Add new samples that are technical replicates (S449). 
```{r}
filename = "/Users/tfriedrich/Downloads/readcounts.txt"
counts1_old = read.table(filename, header = T,stringsAsFactors = F) 
rownames(counts1_old) = counts1_old$Geneid
counts1_gene_annotation = (counts1_old[,1:6])
counts1_old = data.matrix(counts1_old[,7:ncol(counts1_old)])
colnames(counts1_old) = sub("_1_algn_Aligned", "", matrix(unlist(strsplit(colnames(counts1_old), "\\.")), byrow =T, ncol=6)[,3])

filename2 = "/Users/tfriedrich/Downloads/readcounts_newsamples_version3.txt"
counts1_new = read.table(filename2, header=T, stringsAsFactors = F)  
rownames(counts1_new) = counts1_new$Geneid   
counts1_new = data.matrix(counts1_new[,7:ncol(counts1_new)])
colnames(counts1_new) = sub("_1_algn_Aligned", "", matrix(unlist(strsplit(colnames(counts1_new), "\\.")), byrow =T, ncol=7)[,4])

length(which(rownames(counts1_old)==rownames(counts1_new)))

counts1 = data.frame(cbind( counts1_old, counts1_new))      
dim(counts1)

```

Filter out genes with low and high counts ( high counts might be mitochondrial genes)
```{r}
keep = rowMeans(counts1) > 5 & rowMeans(counts1) < 5000 
counts2 = counts1[keep,]
```

Take the log (RPKM)
```{r}
y <- DGEList(counts=counts2)
fpkm_log_matrix = rpkm (y, gene.length=counts1_gene_annotation[keep,"Length"], log = TRUE)
```

Dendrogram 
```{r}
hc2 <- hclust(stats::dist(t(fpkm_log_matrix), method="minkowski"), "ward.D2")
plot(hc2, lwd=4, lty=1, col="black", col.lab="red", xlab="Samples", main="Samples clustered by log2(FPKM+1) using Ward's clustering & Euclidean distance")
```
Some replicates don't cluster (see FocusT replicates). Samples cluster by cell type. SS49 and SS475 are more similar than the other cell types. 

S449 technical replicate showing batch effects. 

Setup up design matrix for PCA plot
```{r}
groupNamesPerSample = c(matrix (unlist(strsplit(colnames(counts2[,1:20]), "_")), byrow=T, ncol=4) [,1], colnames(counts2[,21:24]))
groupNamesPerSample = unlist( lapply (groupNamesPerSample, function(x) sub("S475", "", sub("S449", "", sub("PLC", "", sub("Focus", "", x))))) )
conditionFactor  <- factor(groupNamesPerSample)
subjectgroup = c(matrix (unlist(strsplit(colnames(counts2[,1:20]), "_")), byrow=T, ncol=4) [,1], colnames(counts2[,21:24]))
subjectgroup = unlist( lapply (subjectgroup, function(x) sub("SC", "", sub("YT", "", sub("T", "", sub("Y", "", x))))) )
Subject <- factor(subjectgroup)
```
PCA plot
```{r}
cold <- data.frame("conditionFactor"=conditionFactor, "Subject"=Subject) 
design_matrix <- model.matrix( ~Subject + conditionFactor)
ddsMat <- DESeq2::DESeqDataSetFromMatrix(countData=counts2, colData=cold, design=~conditionFactor);
colnames(ddsMat) <- colnames(counts2)
rld <- DESeq2::rlog(ddsMat) # LOG TRANSFORMED.

```

```{r}
pca              <- prcomp(t(assay(rld)))
allPer.vec       <- 100 * summary(pca)$importance[2,]
allPer.text      <- paste(format(allPer.vec,digits=0, nsmall=1, scientific=FALSE), "%", sep='')
allPerLabels.vec <- paste(names(allPer.vec), "\n", allPer.text, sep='')

barplot(allPer.vec, main="PCA: % variance explained by each component\n" , ylim=c(0,100), names.arg=allPerLabels.vec)

d <- data.frame(pca$x, group=conditionFactor, Subject, name=colnames(rld)) # returns PC1, PC2, ... PC8... etc
ppp <- ggplot(d, aes_string(x="PC1", y="PC2", shape="Subject", color="group"))
ppp <- ppp + geom_point(size=6)
plot(ppp)

ppp <- ggplot(d, aes_string(x="PC1", y="PC2", shape="Subject", color="group"))
ppp <- ppp + geom_point(size=3) + geom_text(aes(label=name), nudge_y=6)
plot(ppp)
```
Samples seem to cluster by cell type. old S449 and S475 appear to be very similar.  Replicates cluster together mostly. 

Group/add up samples with technical replicates 
```{r}
#done manually using counts3 because I don't want to use matrix counts with 1 added to the value
counts3 = data.frame(
            counts2[, "FocusSC_USR17001380L_HJHCJBBXX_L6"]
          , counts2[, "FocusT_USR17001382L_HJHCJBBXX_L7"] + counts2[, "FocusT_USR17001382L_HKF5JBBXX_L1"]
          , counts2[, "FocusYT_USR17001383L_hkff7BBXX_L1"] + counts2[, "FocusYT_USR17001383L_HKY5KBBXX_L1"]
          , counts2[, "FocusY_USR17001381L_HJHCJBBXX_L6"]
          
          , counts2[, "PLCSC_USR17002043L_HJHCJBBXX_L7"]
          , counts2[, "PLCT_USR17002044L_HJHCJBBXX_L7"]
          , counts2[, "PLCYT_USR17001379L_HJHCJBBXX_L6"]
          , counts2[, "PLCY_USR17001377L_HJHCJBBXX_L6"]
          
          , counts2[, "S449SC_USR17001384L_HJHCJBBXX_L7"]
          , counts2[, "S449T_USR17001386L_HJHCJBBXX_L7"] 
          , counts2[, "S449YT_USR17002046L_HJHCJBBXX_L7"] + counts2[, "S449YT_USR17002046L_HKF5JBBXX_L3"]
          , counts2[, "S449Y_USR17001385L_HJHCJBBXX_L7"] + counts2[, "S449Y_USR17001385L_HKF5JBBXX_L3"]
          
          , counts2[, "S475SC_USR17001372L_HJHCJBBXX_L6"]
          , counts2[, "S475T_USR17001374L_HJHCJBBXX_L6"]
          , counts2[, "S475YT_USR17001375L_HJHCJBBXX_L6"]
          , counts2[, "S475Y_USR17001373L_HJHCJBBXX_L6"]
          
          , counts2[, "S449SC"]
          , counts2[, "S449T"]
          , counts2[, "S449Y"]
          , counts2[, "S449YT"]
          )
colnames(counts3)

colnames(counts3) = c(unlist(lapply (strsplit(sub("counts2....", "" ,colnames(counts3[,1:16])) , "_"), `[[`, 1)), sub("counts2.", "", colnames(counts3[, 17:20])))
colnames(counts3) = gsub("\\.", "", colnames(counts3))
rownames(counts3) = rownames(counts2)
```
Look at histogram of counts. 
```{r}
library(ggplot2)
library(tidyr)
ggplot(gather((counts3[,1:4]), cols, value), aes(x = value)) + 
       geom_histogram(binwidth=100) + facet_grid(.~cols)
ggplot(gather((counts3[,5:8]), cols, value), aes(x = value)) + 
       geom_histogram(binwidth=100) + facet_grid(.~cols)
ggplot(gather((counts3[,9:12]), cols, value), aes(x = value)) + 
       geom_histogram(binwidth=100) + facet_grid(.~cols)
ggplot(gather((counts3[,13:16]), cols, value), aes(x = value)) + 
       geom_histogram(binwidth=100) + facet_grid(.~cols)
ggplot(gather((counts3[,17:20]), cols, value), aes(x = value)) + 
       geom_histogram(binwidth=100) + facet_grid(.~cols)
```
Setup up design matrix for Heatmap
```{r}
Condition_heatmap = unlist( lapply (colnames(counts3), function(x) sub("S475", "", sub("S449", "", sub("PLC", "", sub("Focus", "", x))))) )
Condition_heatmap  <- factor(Condition_heatmap)
Subject_heatmap = unlist( lapply (colnames(counts3), function(x) sub("SC", "", sub("YT", "", sub("T", "", sub("Y", "", x))))) )
Subject_heatmap <- factor(Subject_heatmap)
cold <- data.frame("Condition_heatmap"=Condition_heatmap, "Subject_heatmap"=Subject_heatmap) #"sample_name"=colnames(counts4)
#design_matrix2 <- model.matrix( ~Subject2 + Condition)
```
FPKM after combinining samples that are "technical" replicates
```{r}
colnames(counts3) = paste0(colnames(counts3), c(rep("_1", 16), rep("_2", 4)))
y_combined <- DGEList(counts=counts3)
fpkm_matrix = rpkm (y_combined, gene.length=counts1_gene_annotation[keep,"Length"], log = FALSE)
```

Heatmap to see how cells cluster by cell type or experimental condition.  No replicates because they have been combined. 
```{r}
HEATMAP_HEIGHT <- 45 # inchescol
HEATMAP_WIDTH  <- 15
annotation <- data.frame(Expt = Condition_heatmap, Celltype=Subject_heatmap)
rownames(annotation) <- colnames(counts3)
pheatmap(fpkm_matrix,
         , width=HEATMAP_WIDTH, height=HEATMAP_HEIGHT
         , scale="row"  # actually scale each row
         , clustering_distance_rows="correlation" # or: euclidean
         , clustering_distance_cols="correlation"
         , clustering_method="complete"
         , display_numbers=FALSE
         , border_color="#00000000" # note: the last two digits here are the alpha channel/ transparency
         , show_rownames=FALSE, show_colnames = TRUE
         , annotation = annotation
         , fontsize_row=3.5)

```
This heatmap confirms the pattern seen in the dendrogram and PCA plot. 

```{r}
y_combined <- DGEList(counts=counts3)
fpkm_matrix = rpkm (y_combined, gene.length=counts1_gene_annotation[keep,"Length"], log = FALSE)
#wwtr1
fpkm_matrix_wwtr1 = fpkm_matrix[which(rownames(fpkm_matrix) == "ENSG00000018408"), ]
fpkm_matrix_wwtr1_percent = c(cbind(
    fpkm_matrix_wwtr1["FocusT_1"]/fpkm_matrix_wwtr1[ "FocusSC_1"]
  , fpkm_matrix_wwtr1["PLCT_1"]/fpkm_matrix_wwtr1[ "PLCSC_1"]
  , fpkm_matrix_wwtr1["S475T_1"]/fpkm_matrix_wwtr1[ "S475SC_1"]
  , fpkm_matrix_wwtr1["S449T_1"]/fpkm_matrix_wwtr1[ "S449SC_1"]
  , fpkm_matrix_wwtr1["S449T_2"]/fpkm_matrix_wwtr1[ "S449SC_2"]
  , fpkm_matrix_wwtr1["FocusYT_1"]/fpkm_matrix_wwtr1[ "FocusSC_1"]
  , fpkm_matrix_wwtr1["PLCYT_1"]/fpkm_matrix_wwtr1[ "PLCSC_1"]
  , fpkm_matrix_wwtr1["S475YT_1"]/fpkm_matrix_wwtr1[ "S475SC_1"]
  , fpkm_matrix_wwtr1["S449YT_1"]/fpkm_matrix_wwtr1[ "S449SC_1"] 
  , fpkm_matrix_wwtr1["S449YT_2"]/fpkm_matrix_wwtr1[ "S449SC_2"] ))
names(fpkm_matrix_wwtr1_percent ) = c("FocusT/SC", "PLCT/SC", "S475T/SC", "S449T/SC", "S449T_2/SC_2", "FocusYT/SC", "PLCYT/SC", "S475YT/SC", "S449YT/SC", "S449YT_2/SC_2")

#barplot(fpkm_matrix_wwtr1, las=2)
barplot(fpkm_matrix_wwtr1_percent, las=2, cex.names=.7, ylab= "percent KD", main="wwtr1 based on fpkm values")


#yap1
fpkm_matrix_yap1 = fpkm_matrix[which(rownames(fpkm_matrix) == "ENSG00000137693"), ]
fpkm_matrix_yap1_percent = c(cbind(
    fpkm_matrix_yap1["FocusY_1"]/fpkm_matrix_yap1[ "FocusSC_1"]
  , fpkm_matrix_yap1["PLCY_1"]/fpkm_matrix_yap1[ "PLCSC_1"]
  , fpkm_matrix_yap1["S475Y_1"]/fpkm_matrix_yap1[ "S475SC_1"]
  , fpkm_matrix_yap1["S449Y_1"]/fpkm_matrix_yap1[ "S449SC_1"]
  , fpkm_matrix_yap1["S449Y_2"]/fpkm_matrix_yap1[ "S449SC_2"]
  , fpkm_matrix_yap1["FocusYT_1"]/fpkm_matrix_yap1[ "FocusSC_1"]
  , fpkm_matrix_yap1["PLCYT_1"]/fpkm_matrix_yap1[ "PLCSC_1"]
  , fpkm_matrix_yap1["S475YT_1"]/fpkm_matrix_yap1[ "S475SC_1"]
  , fpkm_matrix_yap1["S449YT_1"]/fpkm_matrix_yap1[ "S449SC_1"]
  , fpkm_matrix_yap1["S449YT_2"]/fpkm_matrix_yap1[ "S449SC_2"] ))
names(fpkm_matrix_yap1_percent ) = c("FocusY/SC", "PLCY/SC", "S475Y/SC", "S449Y/SC", "S449Y_2/SC_2",  "FocusYT/SC", "PLCYT/SC", "S475YT/SC", "S449YT/SC", "S449YT_2/SC_2")
#barplot(fpkm_matrix_yap1, las=2)
barplot(fpkm_matrix_yap1_percent, las=2,  cex.names=.7, ylab= "percent KD", main = "yap1 based on fpkm values")

```
S449Y_2 appears to have worked!  Unfortunately, S447T_2 appears to have failed.  It will be more difficult to justify removing samples that didn't knock down its target.  


Remove old S449 samples from matrix before fitting a model. 
Fit the model and estimate dispersion value.
```{r}
counts4 = counts3[, c("FocusSC_1", "FocusT_1", "FocusYT_1", "FocusY_1", "PLCSC_1", "PLCT_1", "PLCYT_1", "PLCY_1", "S475SC_1", "S475T_1", "S475YT_1", "S475Y_1", "S449SC_2", "S449T_2", "S449Y_2", "S449YT_2" )]
```
Setup up design matrix for Heatmap
```{r}
Condition_DE_ = unlist( lapply (matrix(unlist(strsplit(colnames(counts4), "_")), byrow=T, ncol=2)[,1], function(x) sub("S475", "", sub("S449", "", sub("PLC", "", sub("Focus", "", x))))) )
Condition_DE_  <- factor(Condition_DE_)
Subject_DE_ = unlist( lapply (matrix(unlist(strsplit(colnames(counts4), "_")), byrow=T, ncol=2)[,1], function(x) sub("SC", "", sub("YT", "", sub("T", "", sub("Y", "", x))))) )
Subject_DE_ <- factor(Subject_DE_)
design_matrix_DE <- model.matrix( ~Subject_DE_ + Condition_DE_)
colnames(design_matrix_DE)
```

```{r}
y_for_DE <- DGEList(counts=counts4)
y_for_DE <- calcNormFactors(y_for_DE)
y_for_DE = estimateDisp(y_for_DE, design=design_matrix_DE) 
plotBCV(y_for_DE)
fit <- glmFit(y_for_DE, design=design_matrix_DE, dispersion=y_for_DE$trended.dispersion)
```
Run differential expression
SC vs YT
```{r}
lrt_T <- glmLRT(fit,coef = "Condition_DE_T")
topTags(lrt_T)
smoothScatter(x=lrt_T$table$logFC, y=-log10(lrt_T$table$PValue))
lrt_T_table = data.frame(lrt_T$table, p.adjust(lrt_T$table$PValue, method = "BH"))
colnames(lrt_T_table) = c(colnames(lrt_T$table), "adjPVal_BH")
hist(lrt_T_table$PValue)
```
SC vs Y
```{r}
lrt_Y <- glmLRT(fit,coef = "Condition_DE_Y")
topTags(lrt_Y)
smoothScatter(x=lrt_Y$table$logFC, y=-log10(lrt_Y$table$PValue))
lrt_Y_table = data.frame(lrt_Y$table, p.adjust(lrt_Y$table$PValue, method = "BH"))
colnames(lrt_Y_table) = c(colnames(lrt_Y$table), "adjPVal_BH")
hist(lrt_Y_table$PValue)

```
SC vs YT
```{r}
lrt_YT <- glmLRT(fit,coef = "Condition_DE_YT")
topTags(lrt_YT)
smoothScatter(x=lrt_YT$table$logFC, y=-log10(lrt_YT$table$PValue))
lrt_YT_table = data.frame(lrt_YT$table, p.adjust(lrt_YT$table$PValue, method = "BH"))
colnames(lrt_YT_table) = c(colnames(lrt_YT$table), "adjPVal_BH")
hist(lrt_YT_table$PValue)

```


Format output into excel sheet. 
```{r}
genes = rownames(counts4)
ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
head(listAttributes(ensembl), 30)
hgnc_swissprot <- getBM(attributes=c('ensembl_gene_id','hgnc_symbol','description'),filters = 'ensembl_gene_id',  values = genes , mart = ensembl)

fpkm_log_matrix2 = rpkm (y_for_DE, gene.length=counts1_gene_annotation[keep,"Length"], log = FALSE)
index = match (genes, hgnc_swissprot$ensembl_gene_id)
results = data.frame(hgnc_swissprot[index,], counts4, fpkm_log_matrix2, lrt_T_table, lrt_Y_table, lrt_YT_table)
colnames(results) = c(
                    colnames(hgnc_swissprot[index,])
                  , unlist(lapply (colnames(counts4), function(x) paste(x, "counts", sep="_")))
                  , unlist(lapply (colnames(fpkm_log_matrix2), function(x) paste(x, "fpkm", sep="_")))
                  , unlist(lapply (colnames(lrt_T_table), function(x) paste(x, "_SC_vs_T", sep="_")))
                  , unlist(lapply (colnames(lrt_Y_table), function(x) paste(x, "_SC_vs_Y", sep="_")))
                  , unlist(lapply (colnames(lrt_YT_table), function(x) paste(x, "_SC_vs_YT", sep="_"))) )
curr_date = format(Sys.time(), "%a_%b_%d_%H_hrs_%M_min_%S_sec_%Y")
write.table(results, paste0("~/Downloads/Chen_", curr_date, ".xls"), row.names=F, col.names=T, sep="\t")

```
Heatmap of DE genes.  
```{r}
pheatmap(fpkm_log_matrix2[order(lrt_T_table$PValue, decreasing = F)[1:20],c(colnames(fpkm_log_matrix2)[grep("SC", colnames(fpkm_log_matrix2))],"FocusT_1", "PLCT_1", "S475T_1", "S449T_2")], cluster_rows=FALSE, cluster_cols=FALSE)

pheatmap(fpkm_log_matrix2[order(lrt_Y_table$PValue, decreasing = F)[1:20],c(colnames(fpkm_log_matrix2)[grep("SC", colnames(fpkm_log_matrix2))],"FocusY_1", "PLCY_1", "S475Y_1", "S449Y_2")], cluster_rows=FALSE, cluster_cols=FALSE)

pheatmap(fpkm_log_matrix2[order(lrt_YT_table$PValue, decreasing = F)[1:20],c(colnames(fpkm_log_matrix2)[grep("SC", colnames(fpkm_log_matrix2))],colnames(fpkm_log_matrix2)[grep("YT", colnames(fpkm_log_matrix2))] )], cluster_rows=FALSE, cluster_cols=FALSE)

#try this
mean_values = (fit$coefficients %*% t(design_matrix_DE))
colnames(mean_values) = colnames(counts4)
rownames(mean_values) = rownames(counts4)

pheatmap(mean_values[order(lrt_T_table$PValue, decreasing = F)[1:20],c(colnames(fpkm_log_matrix2)[grep("SC", colnames(fpkm_log_matrix2))],"FocusT_1", "PLCT_1", "S475T_1", "S449T_2")], cluster_rows=FALSE, cluster_cols=FALSE)
pheatmap(mean_values[order(lrt_Y_table$PValue, decreasing = F)[1:20],c(colnames(fpkm_log_matrix2)[grep("SC", colnames(fpkm_log_matrix2))],"FocusY_1", "PLCY_1", "S475Y_1", "S449Y_2")], cluster_rows=FALSE, cluster_cols=FALSE)
pheatmap(mean_values[order(lrt_YT_table$PValue, decreasing = F)[1:20],c(colnames(fpkm_log_matrix2)[grep("SC", colnames(fpkm_log_matrix2))],colnames(fpkm_log_matrix2)[grep("YT", colnames(fpkm_log_matrix2))] )], cluster_rows=FALSE, cluster_cols=TRUE)

#edit column description
write.table(colnames(results), "~/Downloads/column_description.txt", quote=F, col.names=F, row.names=F,sep="\t")
```

